docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,844 @@
+However, this must anyway be part of their normal reconciliation logic and is a general best practice of writing Kubernetes controllers.
+
+#### Shoot health checks/conditions
+
+Some of the existing conditions already contain specific code which shall be simplified as well.
+All of the CRDs described above have a `.status.conditions` field to which the controllers may write relevant health information of their function area.
+Gardener will pick them up and copy them over to the Shoots `.status.conditions`.
+
+#### Reconciliation flow
+
+We are now examining the current Shoot creation/reconciliation flow and describe roughly how it would change when applying this proposal:
+
+| Operation | Description |
+|-----------|-------------|
+| botanist.DeployNamespace | Gardener creates the namespace for the Shoot in the Seed cluster. |
+| botanist.DeployKubeAPIServerService | Gardener creates a Service of type `LoadBalancer` in the Seed.<br>AWS Botanist registers a Mutating Webhook and adds its AWS-specific annotation. |
+| botanist.WaitUntilKubeAPIServerServiceIsReady | Gardener checks the `.status` object of the just created `Service` in the Seed. The contract is that also clouds not supporting load balancers must react on the `Service` object and modify the `.status` to correctly reflect the kube-apiserver's ingress IP. |
+| botanist.DeploySecrets | Gardener creates the secrets/certificates it needs like it is today, but it provides utility functions that can be adopted by Botanists/other controllers to also create further certs/secrets on their own. (We should also add labels to all secrets) |
+| botanist.DeployInternalDomainDNSRecord | Gardener creates a DNS-specific CRD in the Seed, and the responsible DNS-controller picks it up and does its job (see CRD above). |
+| botanist.DeployExternalDomainDNSRecord | Gardener creates a DNS-specific CRD in the Seed, and the responsible DNS-controller picks it up and does its job: (see CRD above). |
+| shootCloudBotanist.DeployInfrastructure | Gardener creates a Infrastructure-specific CRD in the Seed, and the responsible Botanist picks it up and does its job: (see CRD above). |
+| botanist.DeployBackupInfrastructure | Gardener creates a `BackupInfrastructure` resource in the Garden cluster.<br>(The BackupInfrastructure controller creates a BackupInfrastructure-specific CRD in the Seed, and the responsible Botanist picks it up and does its job: (see CRD above).) |
+| botanist.WaitUntilBackupInfrastructureReconciled | Gardener checks the `.status` object of the just created `BackupInfrastructure` resource. |
+| hybridBotanist.DeployETCD | Gardener does only deploy the etcd `StatefulSet` without backup-restore sidecar at all.<br>The cloud-specific Botanist registers a Mutating Webhook and adds the backup-restore sidecar, and it also creates the `Secret` needed by the backup-restore sidecar. |
```

Comment:
What about the co-ordinating script that is the entrypoint in the etcd container in the statefulset. Is the `MutatingWebhook` expected to inject that too?

Same comments about `MutatingWebhooks` v/s *proper* controllers apply.

Reply:
> What about the co-ordinating script that is the entrypoint in the etcd container in the statefulset. Is the MutatingWebhook expected to inject that too?

Yes, whatever is needed shall be added by the webhook.